### PR TITLE
[SERVICES-583] Reset password resource

### PR DESCRIPTION
**Pull request by ArturKlajnerok**
This is resource responsible for starting reset password flow for the user. It's providing temporary password sent via email that can be later used with other resource to change the main password.

 * [x] Create a controller for resource that uses user service to start the reset password flow.
 * [x] Throttle getting temporary password for user to one per 24h
 * [x] Save temporary password to storage

/cc @Wikia/services-team

### CODE REVIEW

**Comment by drsnyder** [comment]
Please provide a description.

File: controllers/users_controller.go [hunk 1]
```diff
@@ -94,6 +96,9 @@ func NewUsersController(
 		"/users/{user_id:[0-9]+}/password",
 		timedCorsHandler(controller, "passwordChange", MuxVarsHandler(controller.changeUserPassword).ToHandlerFunc())).Methods("PUT")
 
+	rtr.HandleFunc("/users/reset_password",
```
Comment (drsnyder) [inline controllers/users_controller.go]: I wander if there isn't a better URI for this. The weird part is that we normally accept a user id on the URI and in this case we only have a username.

What about `POST /users/{username}/password`?

File: controllers/users_controller.go [hunk 2]
```diff
@@ -109,6 +114,18 @@ func (controller *UsersController) GetMetricsClient() *client.Client {
...
+	return controller.config
+}
+
+func (controller *UsersController) GetServer() *osin.Server {
+	return controller.server
+}
+
+func (controller *UsersController) GetRedisStorage() *storage.RedisStorage {
+	return controller.redisStorage
+}
```
Comment (drsnyder) [inline controllers/users_controller.go]: Where are these used? Providing these seems like a potential code smell or tight coupling to another part of the code base.

File: controllers/users_controller.go (hunk 1 above)
Comment (ArturKlajnerok) [inline controllers/users_controller.go]: In this case a username value is a requirement, and resource should be usable only with that parameter. I think we can change it to `POST /users/{username}/password` if this makes it more consistent.

File: controllers/users_controller.go (hunk 2 above)
Comment (ArturKlajnerok) [inline controllers/users_controller.go]: These methods were introduced when we added the support for creating tokens from different controller: https://github.com/Wikia/helios/blob/master/controllers/helpers.go#L99 We can clean that up with introduction of a proper auth service that will encapsulate access token creation and requests to osin.

File: models/user.go [hunk 3]
```diff
@@ -90,6 +90,37 @@ func (user *User) IsValid() bool {
...
+	newUser.PasswordHashed = hash
+	return newUser
+}
+
+func (user *User) SetResetPasswordHashed(hash string) *User {
+	newUser := user
+	newUser.ResetPasswordHashed = hash
+	newUser.ResetPasswordTime = utils.GetCurrentFormatedTime()
+	return newUser
+}
```
Comment (drsnyder) [inline models/user.go]: Do these return copies?

File: service/user.go [hunk 4]
```diff
@@ -98,20 +100,40 @@ func (service *UserServiceValue) ChangePassword(id int64, password string) (bool
...
+		errors.Add(UserIsBlockedError)
+		return false, errors
+	}
+
+	if user.IsTemporaryPasswordThrottled() {
+		errors.Add(ThrottledPasswordReset)
+		return false, errors
+	}
+
 	tempPassword, passwordCreationErrors := service.authService.GenerateTemporaryPassword(common.TemporaryPasswordSize)
```
Comment (drsnyder) [inline service/user.go]: In hindsight, maybe `GenerateTemporaryPassword` should have been in `utils`.

**Comment by drsnyder** [comment]
Minor comments, otw LGTM.

File: wikia/wikia.go [hunk 5]
```diff
@@ -127,11 +134,11 @@ func (client *Client) SendPasswordResetEmail(name, newPassword string) (bool, er
 
 	response := &Response{}
 	_, err := client.callWithToken(endpoint, data, response)
-	if err != nil {
+	if err != nil || response.Result != ResultSuccess {
```
Comment (sqreek) [inline wikia/wikia.go]: Can be this replaced by
`return !(err!=nil || response.Result != ResultSuccess), err`
or it's out of convention?

**Comment by nmonterroso** [comment]
Can you run `./generateResources.sh`? otherwise LGTM

File: controllers/users_controller.go (hunk 2 above)
Comment (ArturKlajnerok) [inline controllers/users_controller.go]: @drsnyder Changed in : 861f2e8

File: models/user.go (hunk 3 above)
Comment (ArturKlajnerok) [inline models/user.go]: Good catch, in this case to make a proper copy pointer needed to be dereferenced, fixed in : 39a784a

File: service/user.go (hunk 4 above)
Comment (ArturKlajnerok) [inline service/user.go]: For mocking purposes it makes sense to keep `GenerateTemporaryPassword` in `AuthService`.

File: wikia/wikia.go (hunk 5 above)
Comment (ArturKlajnerok) [inline wikia/wikia.go]: This should also work, but it's out of current convention.

**Pull request merged**